scipy/special/_iv_ratio.h

```diff
@@ -0,0 +1,96 @@
+// Numerically stable computation of iv(v+1, x) / iv(v, x)
+
+#pragma once
+
+#include "special/tools.h"
+#include "special/error.h"
+#include <cstdint>    // for uint64_t
+#include <cmath>      // for frexp, ldexp, fmax, fma
+#include <utility>    // for pair
+
+/* Generates the "tail" of Perron's continued fraction for `iv(v,x)/iv(v-1,x)`
+ * for v >= 1 and x >= 0.
+ *
+ * The Perron continued fraction is studied in [1].  It is given by
+ *
+ *         iv(v, x)      x    -(2v+1)x   -(2v+3)x   -(2v+5)x
+ *   R := --------- = ------ ---------- ---------- ---------- ...
+ *        iv(v-1,x)   2v+x + 2(v+x)+1 + 2(v+x)+2 + 2(v+x)+3 +
+ *
+ * Rearrange the expression by making an equivalent transform to prevent
+ * floating point overflow and extracting the first fraction to simplify
+ * the recurrence relation.  This leads to
+ *
+ *        xc                -(2vc+c)(xc) -(2vc+3c)(xc) -(2vc+5c)(xc)
+ *   R = -----,  fc = 2vc + ------------ ------------- ------------- ...
+ *       xc+fc              2(vc+xc)+c + 2(vc+xc)+2c + 2(vc+xc)+3c +
+ *
+ * This class generates the fractions of fc after 2vc.
+ *
+ * [1] Gautschi, W. and Slavik, J. (1978). "On the computation of modified
+ *     Bessel function ratios." Mathematics of Computation, 32(143):865-875.
+ */
+struct IvRatioCFTailGenerator {
+
+    // It is assumed that v >= 1, x >= 0, c > 0, and all are finite.
+    IvRatioCFTailGenerator(double vc, double xc, double c) noexcept {
+        a0_ = -(2*vc-c)*xc;
+        as_ = -2*c*xc;
+        b0_ = 2*(vc+xc);
+        bs_ = c;
+        k_ = 0;
+    }
+
+    std::pair<double, double> operator()() {
+        ++k_;
+        return {std::fma(k_, as_, a0_), std::fma(k_, bs_, b0_)};
+    }
+
+private:
+    double a0_, as_;  // a[k] == a0 + as*k, k >= 1
+    double b0_, bs_;  // b[k] == b0 + bs*k, k >= 1
+    std::uint64_t k_; // current index
+};
+
+inline double iv_ratio(double v, double x, int comp) {
```

Comment:
Do you anticipate passing arrays for `comp` in the `ufunc`? For other cases in `special`, the ufunc for the complement is a separate function, and I think that's what we should do here to.  I like the way template metaprogramming is used here: https://github.com/scipy/scipy/blob/1457f69ee3acf1093731e724132729b23a0fc3d7/scipy/special/special/wright_bessel.h#L659-L660

to write two very similar functions with slightly different behavior based on a boolean switch.

I guess `iv_ratio_c` would be a reasonable name for the complement for now. We've been discussing a naming overhaul though, so this would likely be changed somewhere down the line.

Reply:
Good idea, a separate function is clearer!

I removed the “comp” argument now to keep things simple for this PR. Another reason is that the relative truncation error of `iv_ratio` is provably small (a few machine epsilon), but the same simple bound doesn’t apply to its complement.

I’ll spend more time to work on the truncation bound of `iv_ratio_c`, and attempt it in a separate PR later, possibly together with `iv_ratio_inv`.